### PR TITLE
Fix InstructPix2Pix training in multi-GPU mode

### PR DESCRIPTION
**Pull request by sayakpaul**
Should close https://github.com/huggingface/diffusers/issues/2966.

Command used for testing:

```bash

accelerate launch --mixed_precision="fp16" --multi_gpu train_instruct_pix2pix.py \
 --pretrained_model_name_or_path=runwayml/stable-diffusion-v1-5 \
 --dataset_name=sayakpaul/instructpix2pix-1000-samples \
 --use_ema \
 --enable_xformers_memory_efficient_attention \
 --resolution=512 --random_flip \
 --train_batch_size=4 --gradient_accumulation_steps=4 --gradient_checkpointing \
 --max_train_steps=100 \
 --checkpointing_steps=10 --checkpoints_total_limit=1 \
 --learning_rate=5e-05 --lr_warmup_steps=0 \
 --conditioning_dropout_prob=0.05 \
 --mixed_precision=fp16 \
 --seed=42 
```

### CODE REVIEW

**Review by patrickvonplaten — APPROVED** [review]
Looks good!

**Comment by sayakpaul** [comment]
As pointed out by @whbzju [here](https://github.com/huggingface/diffusers/issues/2966#issuecomment-1498501690), we also need to do unwrapping while running validation inference. So, I fixed that and did the testing with the following command: 

```bash
accelerate launch --mixed_precision="fp16" --multi_gpu train_instruct_pix2pix.py \
--pretrained_model_name_or_path=runwayml/stable-diffusion-v1-5 \
--dataset_name=sayakpaul/instructpix2pix-1000-samples \
--use_ema \
--enable_xformers_memory_efficient_attention \
--resolution=512 --random_flip \
--train_batch_size=2 --gradient_accumulation_steps=4 --gradient_checkpointing \
--max_train_steps=20 \
--checkpointing_steps=10 --checkpoints_total_limit=1 \
--learning_rate=5e-05 --lr_warmup_steps=0 \
--conditioning_dropout_prob=0.05 \
--mixed_precision=fp16 \
--val_image_url="https://hf.co/datasets/diffusers/diffusers-images-docs/resolve/main/mountain.png" \
--validation_prompt="make the mountains snowy" \
--seed=42 \
--report_to=wandb
```

@patrickvonplaten could you take a look again?

File: examples/instruct_pix2pix/train_instruct_pix2pix.py [hunk 1]
```diff
@@ -904,7 +906,9 @@ def collate_fn(examples):
                 # run inference
                 original_image = download_image(args.val_image_url)
                 edited_images = []
-                with torch.autocast(str(accelerator.device), enabled=accelerator.mixed_precision == "fp16"):
+                with torch.autocast(
+                    str(accelerator.device).replace(":0", ""), enabled=accelerator.mixed_precision == "fp16"
+                ):
```
Comment (williamberman) [inline examples/instruct_pix2pix/train_instruct_pix2pix.py:911]:
```suggestion
                with torch.autocast(
                    "cuda", enabled=accelerator.mixed_precision == "fp16"
                ):
```

We generally just directly pass "cuda" in our other training scripts. I think we're generally open to our training scripts explicitly targeting cuda and not being device agnostic -- i.e. later in the function we explicitly call torch.cuda.empty_cache

File: examples/instruct_pix2pix/train_instruct_pix2pix.py [hunk 2]
```diff
@@ -959,7 +963,7 @@ def collate_fn(examples):
         if args.validation_prompt is not None:
             edited_images = []
             pipeline = pipeline.to(accelerator.device)
-            with torch.autocast(str(accelerator.device)):
+            with torch.autocast(str(accelerator.device).replace(":0", "")):
```
Comment (williamberman) [inline examples/instruct_pix2pix/train_instruct_pix2pix.py:966]:
```suggestion
            with torch.autocast("cuda"):
```

**Review by williamberman — APPROVED** [review]

File: examples/instruct_pix2pix/train_instruct_pix2pix.py (hunk 1 above)
Comment (sayakpaul) [reply examples/instruct_pix2pix/train_instruct_pix2pix.py:911]: Good point!

**Pull request merged**